### ISSUE TITLE
StackOverflowError because of cyclic DETACH

### ISSUE THREAD

**Issue by mheydem** [eclipse-ee4j/eclipselink#1577]
After upgrading from
- Java SE 8
- Glassfish 5.1
- Jakarta EE 8
- EclipseLink 2.7.4

to
- Java SE 8
- Glassfish 6.0
- Jakarta EE 9.0
- EclipseLink 3.0.0

the action
- detaching an entity which has a bidirectional oneToMany relation

causes
- cyclic detaching ending up in an StackOverflowError
- DescriptorIterator.iterateIndirectContainerForMapping loops until StackOverflowError

Here an example:
```
public class Group {

	@Column(name = "NAME")
	@NotNull
	private String name;

	@OneToMany(
		cascade = {CascadeType.ALL },
		fetch = FetchType.LAZY,
		mappedBy = "group",
		targetEntity = GroupItem.class)
	@MapKey(name = "name")
	@NotNull
	private Map<String, GroupItem<?>> properties = new HashMap<>();
}
```

```
public class GroupItem {

	@ManyToOne(
		cascade = { CascadeType.ALL },
		fetch = FetchType.EAGER,
		optional = false)
	@JoinColumn(name = "GROUP_ID")
 	@NotNull
 	private Group group;
}
```
The code above is working fine with EclipseLink 2.x, but when running on EclipseLink 3.0.2  and detaching a `Group` is tried a StackOverflowError can be experienced.

Removing the `CascadeType.Detach` from one side of the relation avoids the StackOverflowError, but is not a real workaround as it will not behave exactly the same way:

```
public class GroupItem {
	@ManyToOne(
			cascade = {
					CascadeType.PERSIST,
					CascadeType.MERGE,
					CascadeType.REMOVE,
					CascadeType.REFRESH
					/*CascadeType.DETACH*/ },
			fetch = FetchType.EAGER,
			optional = false)
	@JoinColumn(name = "GROUP_ID")
	@NotNull
	private Group group;
```

Here the stack trace:
```
Caused by: java.lang.StackOverflowError
	at sun.nio.ch.IOUtil.write(IOUtil.java:58)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:470)
	at oracle.net.nt.TimeoutSocketChannel.write(TimeoutSocketChannel.java:215)
	at oracle.net.ns.NIOPacket.writeToSocketChannel(NIOPacket.java:211)
	at oracle.net.ns.NIONSDataChannel.writeDataToSocketChannel(NIONSDataChannel.java:181)
	at oracle.net.ns.NIONSDataChannel.writeDataToSocketChannel(NIONSDataChannel.java:132)
	at oracle.jdbc.driver.T4CMAREngineNIO.prepareForReading(T4CMAREngineNIO.java:96)
	at oracle.jdbc.driver.T4CMAREngineNIO.unmarshalUB1(T4CMAREngineNIO.java:534)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:485)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:252)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:612)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:226)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:59)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:747)
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:904)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1082)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3780)
	at oracle.jdbc.driver.T4CPreparedStatement.executeInternal(T4CPreparedStatement.java:1343)
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:3822)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1165)
	at com.sun.gjc.spi.jdbc40.PreparedStatementWrapper40.executeQuery(PreparedStatementWrapper40.java:618)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeSelect(DatabaseAccessor.java:1015)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:651)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeCall(DatabaseAccessor.java:567)
	at org.eclipse.persistence.internal.sessions.AbstractSession.basicExecuteCall(AbstractSession.java:2045)
	at org.eclipse.persistence.sessions.server.ClientSession.executeCall(ClientSession.java:312)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:282)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:268)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.selectOneRow(DatasourceCallQueryMechanism.java:794)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectOneRowFromTable(ExpressionQueryMechanism.java:2903)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectOneRow(ExpressionQueryMechanism.java:2856)
	at org.eclipse.persistence.queries.ReadObjectQuery.executeObjectLevelReadQuery(ReadObjectQuery.java:563)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:912)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadObjectQuery.execute(ReadObjectQuery.java:448)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2984)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1844)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1826)
	at org.eclipse.persistence.internal.indirection.NoIndirectionPolicy.valueFromQuery(NoIndirectionPolicy.java:351)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRowInternal(ForeignReferenceMapping.java:2344)
	at org.eclipse.persistence.mappings.OneToOneMapping.valueFromRowInternal(OneToOneMapping.java:1883)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.valueFromRow(ForeignReferenceMapping.java:2185)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.buildCloneFromRow(ForeignReferenceMapping.java:341)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildAttributesIntoWorkingCopyClone(ObjectBuilder.java:2063)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildWorkingCopyCloneFromRow(ObjectBuilder.java:2316)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObjectInUnitOfWork(ObjectBuilder.java:914)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:800)
	at org.eclipse.persistence.internal.descriptors.ObjectBuilder.buildObject(ObjectBuilder.java:754)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.buildObject(ObjectLevelReadQuery.java:863)
	at org.eclipse.persistence.queries.ReadAllQuery.registerResultInUnitOfWork(ReadAllQuery.java:987)
	at org.eclipse.persistence.queries.ReadAllQuery.executeObjectLevelReadQuery(ReadAllQuery.java:598)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:912)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadAllQuery.execute(ReadAllQuery.java:485)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2984)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1844)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1826)
	at org.eclipse.persistence.internal.indirection.QueryBasedValueHolder.instantiate(QueryBasedValueHolder.java:136)
	at org.eclipse.persistence.internal.indirection.QueryBasedValueHolder.instantiateForUnitOfWorkValueHolder(QueryBasedValueHolder.java:155)
	at org.eclipse.persistence.internal.indirection.UnitOfWorkValueHolder.instantiateImpl(UnitOfWorkValueHolder.java:163)
	at org.eclipse.persistence.internal.indirection.UnitOfWorkValueHolder.instantiate(UnitOfWorkValueHolder.java:242)
	at org.eclipse.persistence.internal.indirection.DatabaseValueHolder.getValue(DatabaseValueHolder.java:97)
	at org.eclipse.persistence.indirection.IndirectMap.buildDelegate(IndirectMap.java:129)
	at org.eclipse.persistence.indirection.IndirectMap.getDelegate(IndirectMap.java:416)
	at org.eclipse.persistence.indirection.IndirectMap$1.<init>(IndirectMap.java:226)
	at org.eclipse.persistence.indirection.IndirectMap.entrySet(IndirectMap.java:225)
	at org.eclipse.persistence.internal.queries.MapContainerPolicy$MapContainerPolicyIterator.<init>(MapContainerPolicy.java:886)
	at org.eclipse.persistence.internal.queries.MapContainerPolicy.iteratorFor(MapContainerPolicy.java:554)
	at org.eclipse.persistence.mappings.CollectionMapping.iterateOnRealAttributeValue(CollectionMapping.java:1394)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.iterateIndirectContainerForMapping(DescriptorIterator.java:276)
	at org.eclipse.persistence.internal.indirection.TransparentIndirectionPolicy.iterateOnAttributeValue(TransparentIndirectionPolicy.java:507)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.iterate(ForeignReferenceMapping.java:1457)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.internalIterateReferenceObjects(DescriptorIterator.java:444)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.iterateReferenceObjects(DescriptorIterator.java:397)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.iterateReferenceObjectForMapping(DescriptorIterator.java:377)
	at org.eclipse.persistence.mappings.ObjectReferenceMapping.iterateOnRealAttributeValue(ObjectReferenceMapping.java:379)
	at org.eclipse.persistence.internal.indirection.IndirectionPolicy.iterateOnAttributeValue(IndirectionPolicy.java:282)
	at org.eclipse.persistence.mappings.ForeignReferenceMapping.iterate(ForeignReferenceMapping.java:1457)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.internalIterateReferenceObjects(DescriptorIterator.java:444)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.iterateReferenceObjects(DescriptorIterator.java:397)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.iterateReferenceObjectForMapping(DescriptorIterator.java:377)
	at org.eclipse.persistence.mappings.CollectionMapping.iterateOnElement(CollectionMapping.java:1380)
	at org.eclipse.persistence.mappings.CollectionMapping.iterateOnRealAttributeValue(CollectionMapping.java:1397)
	at org.eclipse.persistence.internal.descriptors.DescriptorIterator.iterateIndirectContainerForMapping(DescriptorIterator.java:276)
... looping ...
```

**Comment by mheydem** [comment]
Adapted bug report as applies for all CascadeTypes.

**Comment by rfelcman** [comment]
Hello,
I tried to simulate this issue but without success now (no crash). I have some additional questions (see `TODO question` sequences in the attached test case - Maven project):

- About dependencies. Which validation API and implementation is used?
- About persistence/EclipseLink logging. Which log level is specified?
- This should be related with logging too. Is and how is implemented toString() method in the entities?
- JPA operations which lead to the crash. Which JPA operations preceding to the crash?
- Entities. Generic GroupItem<?> is used. How it's declared in GroupItem class?

In the attached test case is _init.sql_ script to create required tables, java code (entities, test) and _persistence.xml_ file. Feel free to modify it and provide it back. This is easiest way to simulate this issue -> fix it.

[jpa-bug-1577-StackOverflowErrorBecauseOfCyclicDETACH.tar.gz](https://github.com/eclipse-ee4j/eclipselink/files/9431861/jpa-bug-1577-StackOverflowErrorBecauseOfCyclicDETACH.tar.gz)

**Comment by mheydem** [comment]
@rfelcman Thank you for investigating and providing the test project.

**Finally**
Finally I removed the _detach_ command from our productive code as it is not really necessary at that place.
I couldn't find the reason for the _StackOverflowError_.
There's a bug, but it seems to show up under certain circumstances only.

**Productive Code**
More testing and debugging in the productive code with finer log levels for EclipseLink was done.
Unfortunately without further knowledge.
Finally it always ends up in debugging the loop.

**Test Project**
I adapted the test project to replicate all the relevant parts from productive code and took care about the questions you rised above. Unfortunately the issue was not reproducable in the test project.

**Comment by mheydem** [comment]
We have now progressed further in the Jakarta EE9 migration and have also been able to carry out a larger amount of tests. More code passages with calls to detach have been noticed that trigger this bug. Unfortunately, I can only create the problem in our complex large setup.
The last comment by @ethael sounds like a further migration to Jakarta EE 10 will not solve the problem.

**Comment by mheydem** [comment]
When I change the FetchType of the properties in Group form LAZY to EAGER in my setup there is no StackOverflowException. Unfortunately I am still not able reproduce this in a small and simple example.
However, I hope that this additional information will somehow help to solve this problem.

**Comment by mister48** [comment]
I modified the sample project to get the error:
- Add the `GroupItem` to the properties of `Group`
- Fetch `Group` only
- Detach `Group`

In my opinion the PR #348 introduced the change in the `DescriptorIterator`. My understanding of the bug is as follows:
1. detach Group
2. fetch lazy GroupItem from DB
3. detach GroupItem
4. detach eager Group (same entity, not same object, but visited objects are tracked with IdentityHashMap)
5. fetch lazy GroupItem from DB (again same entity, not same object, therefore again not in visited objects)
6. etc.

[jpa-bug-1577-StackOverflowErrorBecauseOfCyclicDETACH.tar.gz](https://github.com/eclipse-ee4j/eclipselink/files/11674295/jpa-bug-1577-StackOverflowErrorBecauseOfCyclicDETACH.tar.gz)

**Comment by rfelcman** [comment]
Fixed by  #2037
